mod-infra/src/main/scala/at/forsyte/apalache/infra/passes/Pass.scala

```diff
@@ -37,4 +39,6 @@ trait Pass {
    * @return the next pass, if exists, or None otherwise
    */
   def next(): Option[Pass]
+
+  def dependencies: Set[ModuleProperty.Value]
```

Comment:
Yeah, I'm still wondering whether it should include it or not. It makes sense that it should, but that would imply that every pass depends on parsing, for example. What's your opinion on this?

Reply:
I can see the merits of both approaches, but my personal preference is to avoid implicit transitivity in these cases. For the particular case of parsing, you can see my comment on that, but in general, if a pass is a prerequisite for literally everything, there is no value in tracking that info with a variable (since it'd be in literally every set). 
As for the reason I prefer explicitly listing all dependencies: it forces us (well, you for now, I guess) to understand and write down _why_ one pass requires a certain set of properties. For instance, why does normalization require `TypeChecking`? If you think about it, all normal form transformations are type-preserving, so it's not actually necessary to type-check beforehand. On the other hand, simplification has a critical dependency on typechecking. Consider the following:
```tla
\* @type: (Int, Bool) => Int
A(p,q) == IF q THEN 0 ELSE p + 1

B == A("abc", TRUE) > 0
```
Assuming no typechecking, inlining transforms this into
```tla
B == IF TRUE THEN 0 ELSE "abc" + 1
```
so far, not a problem by itself, because running typechecking after inlining would reveal the exact same type-mismatch as running it before: `"abc"` passed to a function expecting an `Int`. However, if simplification is run after inlining but before typechecking, we get 
```tla
B == 0
```
which hides the error. To me, this is basically what it boils down to: a property should be in a pass' `dependencies` iff there exists an error that is not caught when the pass that ensures the property is run later. That's why I wouldn't consider typechecking to be a dependency of inlining, for instance, because the error still gets caught, just later in the pipeline.